driver/avgpool_driver.hpp

```diff
@@ -0,0 +1,589 @@
+    int ParseCmdLineArgs(int argc, char* argv[]) override;
+    InputFlags& GetInputFlags() override { return inflags; }
+
+    int GetandSetData() override;
+
+    int AllocateBuffersAndCopy() override;
+
+    int RunForwardGPU() override;
+    int RunForwardCPU();
+
+    int RunBackwardGPU() override;
+    int RunBackwardCPU();
+
+    Tref GetTolerance();
+    int VerifyBackward() override;
+    int VerifyForward() override;
+    ~AvgPoolDriver() override
+    {
+        miopenDestroyTensorDescriptor(inputDesc);
+        miopenDestroyTensorDescriptor(outputDesc);
+        miopenDestroyTensorDescriptor(inputGradDesc);
+        miopenDestroyTensorDescriptor(outputGradDesc);
+    }
+
+private:
+    InputFlags inflags;
+
+    int forw;
+
+    miopenTensorDescriptor_t inputDesc;
+    miopenTensorDescriptor_t outputDesc;
+    miopenTensorDescriptor_t inputGradDesc;
+    miopenTensorDescriptor_t outputGradDesc;
+
+    std::unique_ptr<GPUMem> input_dev;
+    std::unique_ptr<GPUMem> output_dev;
+    std::unique_ptr<GPUMem> input_grad_dev;
+    std::unique_ptr<GPUMem> output_grad_dev;
+
+    std::vector<Tgpu> input;
+    std::vector<Tgpu> output;
+    std::vector<Tref> output_host;
+    std::vector<Tgpu> input_grad;
+    std::vector<Tref> input_grad_host;
+    std::vector<Tgpu> output_grad;
+    std::vector<int64_t> ksize;
+    std::vector<int64_t> stride;
+    std::vector<int64_t> padding;
+
+    bool ceil_mode;
+    bool count_include_pad;
+    int64_t divisor_override;
+    int64_t N, C, D, H, W, OD, OH, OW;
+
+    std::vector<uint64_t> in_dim;
+    bool isContiguous;
+};
+
+template <typename Tgpu, typename Tref>
+int AvgPoolDriver<Tgpu, Tref>::ParseCmdLineArgs(int argc, char* argv[])
+{
+    inflags.Parse(argc, argv);
+    isContiguous = inflags.GetValueInt("is-contiguous") == 1 ? true : false;
+
+    if(inflags.GetValueInt("time") == 1)
+    {
+        miopenEnableProfiling(GetHandle(), true);
+    }
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int AvgPoolDriver<Tgpu, Tref>::GetandSetData()
+{
+    in_dim                            = inflags.GetValueTensorUint64("input_dims").lengths;
+    std::vector<uint64_t> in_stride   = ComputeStrides(in_dim);
+    int ksp_dim                       = in_dim.size() - 2;
+    std::vector<uint64_t> ksize_int   = inflags.GetValueTensorUint64("kernel_size").lengths;
+    ksize                             = std::vector<int64_t>(ksize_int.begin(), ksize_int.end());
+    std::vector<uint64_t> stride_int  = inflags.GetValueTensorUint64("stride").lengths;
+    stride                            = std::vector<int64_t>(stride_int.begin(), stride_int.end());
+    std::vector<uint64_t> padding_int = inflags.GetValueTensorUint64("padding").lengths;
+    padding = std::vector<int64_t>(padding_int.begin(), padding_int.end());
+
+    if(ksize.size() != ksp_dim)
+    {
+        int ref = ksp_dim - ksize.size();
+        while((ref--) != 0)
+            ksize.push_back(ksize[0]);
+    }
```

Comment:
Just want to clarify: does this code try to duplicate the number to create kernel size when ksize.size() != in_dim.size() - 2 ? Does that mean when ksize.size() > ksp_dim the code will be infinite loop?

Reply:
add `if(ref < 0)
            MIOPEN_THROW("Invalid kernel size");`